Userland/Libraries/LibWeb/Layout/InitialContainingBlock.cpp

```diff
@@ -58,38 +64,31 @@ void InitialContainingBlock::recompute_selection_states()
 {
     SelectionState state = SelectionState::None;
 
-    auto selection = this->selection().normalized();
+    auto selection = document().get_selection();
+    auto range = selection ? selection->range() : nullptr;
 
+    // FIXME: Make sure this works with consecutive non-text children of things.
     for_each_in_inclusive_subtree([&](auto& layout_node) {
-        if (!selection.is_valid()) {
+        auto* dom_node = layout_node.dom_node();
+        if (!range) {
             // Everything gets SelectionState::None.
-        } else if (&layout_node == selection.start().layout_node && &layout_node == selection.end().layout_node) {
+        } else if (dom_node && dom_node == range->start_container() && dom_node == range->end_container()) {
             state = SelectionState::StartAndEnd;
-        } else if (&layout_node == selection.start().layout_node) {
+        } else if (dom_node && dom_node == range->start_container()) {
             state = SelectionState::Start;
-        } else if (&layout_node == selection.end().layout_node) {
+        } else if (dom_node && dom_node == range->end_container()) {
             state = SelectionState::End;
         } else {
-            if (state == SelectionState::Start)
+            if (state == SelectionState::Start || (state == SelectionState::StartAndEnd && dom_node && !is<DOM::Text>(*dom_node))) {
                 state = SelectionState::Full;
-            else if (state == SelectionState::End || state == SelectionState::StartAndEnd)
+            }
+            else if (state == SelectionState::End || state == SelectionState::StartAndEnd) {
                 state = SelectionState::None;
+            }
```

Comment:
This part (added with a fixup) is not formatted correctly, it should be `} else if (...) {`.
The curlies could also be removed since they were not present originally, redicing the delta from `2-4+` to just `1+1-`.